python/lsst/fgcmcal/fgcmBuildStarsTable.py

```diff
@@ -0,0 +1,347 @@
+            raise RuntimeError("Must set calibFluxApertureRadius if doSubtractLocalBackground is True.")
+
+        # To get the correct output schema, we use the same code as fgcmBuildStarsTask
+        # We are not actually using this mapper, except to grab the outputSchema
+        dataRef = groupedDataRefs[list(groupedDataRefs.keys())[0]][0]
+        sourceSchema = dataRef.get('src_schema', immediate=True).schema
+        sourceMapper = self._makeSourceMapper(sourceSchema)
+        outputSchema = sourceMapper.getOutputSchema()
+
+        # Construct mapping from ccd number to index
+        camera = dataRef.get('camera')
+        ccdMapping = {}
+        for ccdIndex, detector in enumerate(camera):
+            ccdMapping[detector.getId()] = ccdIndex
+
+        approxPixelAreaFields = computeApproxPixelAreaFields(camera)
+
+        if inStarObsCat is not None:
+            fullCatalog = inStarObsCat
+            comp1 = fullCatalog.schema.compare(outputSchema, outputSchema.EQUAL_KEYS)
+            comp2 = fullCatalog.schema.compare(outputSchema, outputSchema.EQUAL_NAMES)
+            if not comp1 or not comp2:
+                raise RuntimeError("Existing fgcmStarObservations file found with mismatched schema.")
+        else:
+            fullCatalog = afwTable.BaseCatalog(outputSchema)
+
+        visitKey = outputSchema['visit'].asKey()
+        ccdKey = outputSchema['ccd'].asKey()
+        instMagKey = outputSchema['instMag'].asKey()
+        instMagErrKey = outputSchema['instMagErr'].asKey()
+
+        # Prepare local background if desired
+        if self.config.doSubtractLocalBackground:
+            # localBackgroundFluxKey = sourceSchema[self.config.localBackgroundFluxField].asKey()
+            localBackground = 0.0
+            # localBackgroundArea = np.pi*calibFluxApertureRadius**2.
+        else:
+            localBackground = 0.0
+
+        # Determine which columns we need from the sourceTable_visit catalogs
+        columns = ['ra', 'decl', 'x', 'y', self.config.psfCandidateName,
+                   self.config.instFluxField, self.config.instFluxField + 'Err',
+                   self.config.apertureInnerInstFluxField, self.config.apertureInnerInstFluxField + 'Err',
+                   self.config.apertureOuterInstFluxField, self.config.apertureOuterInstFluxField + 'Err']
+        if self.sourceSelector.doFlags:
+            columns.extend(self.sourceSelector.flags.bad)
+        if self.sourceSelector.doUnresolved:
+            columns.append(self.sourceSelector.unresolved.name)
+        if self.sourceSelector.doIsolated:
+            columns.append(self.sourceSelector.isolated.parentName)
+            columns.append(self.sourceSelector.isolated.nChiledName)
+        if self.config.doSubtractLocalBackground:
+            columns.append(self.config.localBackgroundFluxField)
+
+        k = 2.5/np.log(10.)
+
+        # loop over visits
+        for ctr, visit in enumerate(visitCat):
```

Comment:
It's a bunch of duplicated math, and two completely different APIs unfortunately.  (For example, to access the values of a pandas dataframe you need `.values` while that doesn't make sense for numpy/afwTable).  We'll end up deprecating the gen2/src access so I thought that jumping through hoops to consolidate these didn't make sense.

Reply:
I do still feel the logic would be much clearer if the loop was a separate method, e.g. `process_visit(visit)`. This is a very long method and for loop, otherwise.